Data/HashMap/Internal.hs

```diff
@@ -1771,37 +1756,147 @@ differenceWith f a b = foldlWithKey' go empty a
+    -- collision vs. collision
+    go _ (Collision h1 ls1) (Collision h2 ls2)
+      | h1 == h2 = runST $ do
+        (len, mary) <- intersectionCollisions f ls1 ls2
+        case len of
+          0 -> pure Empty
+          1 -> Leaf h1 <$> A.read mary 0
+          _ -> Collision h1 <$> (A.unsafeFreeze =<< A.shrink mary len)
+      | otherwise = Empty
+    -- branch vs. branch
+    go s (BitmapIndexed b1 ary1) (BitmapIndexed b2 ary2) = intersectionArray s b1 b2 ary1 ary2
+    go s (BitmapIndexed b1 ary1) (Full ary2) = intersectionArray s b1 fullNodeMask ary1 ary2
+    go s (Full ary1) (BitmapIndexed b2 ary2) = intersectionArray s fullNodeMask b2 ary1 ary2
+    go s (Full ary1) (Full ary2) = intersectionArray s fullNodeMask fullNodeMask ary1 ary2
+    -- collision vs. branch
+    go s (BitmapIndexed b1 ary1) t2@(Collision h2 _ls2)
+      | b1 .&. m2 == 0 = Empty
+      | otherwise = go (s + bitsPerSubkey) (A.index ary1 i) t2
+      where
+        m2 = mask h2 s
+        i = sparseIndex b1 m2
+    go s t1@(Collision h1 _ls1) (BitmapIndexed b2 ary2)
+      | b2 .&. m1 == 0 = Empty
+      | otherwise = go (s + bitsPerSubkey) t1 (A.index ary2 i)
+      where
+        m1 = mask h1 s
+        i = sparseIndex b2 m1
+    go s (Full ary1) t2@(Collision h2 _ls2) = go (s + bitsPerSubkey) (A.index ary1 i) t2
+      where
+        i = index h2 s
+    go s t1@(Collision h1 _ls1) (Full ary2) = go (s + bitsPerSubkey) t1 (A.index ary2 i)
+      where
+        i = index h1 s
+
+    intersectionArray s b1 b2 ary1 ary2
+      -- don't create an array of size zero in intersectionArrayBy
+      | b1 .&. b2 == 0 = Empty
+      | otherwise = runST $ do
+        (b, len, ary) <- intersectionArrayBy (go (s + bitsPerSubkey)) b1 b2 ary1 ary2
+        case len of
+          0 -> pure Empty
+          1 -> A.read ary 0
+          _ -> bitmapIndexedOrFull b <$> (A.unsafeFreeze =<< A.shrink ary len)
+{-# INLINE intersectionWithKey# #-}
+
+intersectionArrayBy ::
+  ( HashMap k v1 ->
+    HashMap k v2 ->
+    HashMap k v3
+  ) ->
+  Bitmap ->
+  Bitmap ->
+  A.Array (HashMap k v1) ->
+  A.Array (HashMap k v2) ->
+  ST s (Bitmap, Int, A.MArray s (HashMap k v3))
+intersectionArrayBy f !b1 !b2 !ary1 !ary2 = do
+  mary <- A.new_ $ popCount bIntersect
+  -- iterate over nonzero bits of b1 .&. b2
+  let go !i !i1 !i2 !b !bFinal
+        | b == 0 = pure (i, bFinal)
+        | testBit $ b1 .&. b2 = do
+          x1 <- A.indexM ary1 i1
+          x2 <- A.indexM ary2 i2
+          case f x1 x2 of
+            Empty -> go i (i1 + 1) (i2 + 1) b' (bFinal .&. complement m)
+            _ -> do
+              A.write mary i $! f x1 x2
+              go (i + 1) (i1 + 1) (i2 + 1) b' bFinal
+        | testBit b1 = go i (i1 + 1) i2 b' bFinal
+        | otherwise = go i i1 (i2 + 1) b' bFinal
+        where
+          m = 1 `unsafeShiftL` countTrailingZeros b
+          testBit x = x .&. m /= 0
+          b' = b .&. complement m
+  (maryLen, bFinal) <- go 0 0 0 bCombined bIntersect
```

Comment:
The comment at the top seems incorrect: Currently the loop actually iterates over `b1 .|. b2`. It would be nice to change this though. In that case the `i1` and `i2` indices could be computed with `sparseIndex`.

Reply:
> In that case the i1 and i2 indices could be computed with sparseIndex.

For comparison, here is a version of  `unionArrayBy` that uses `sparseIndex` to compute all the indices:

https://github.com/haskell-unordered-containers/unordered-containers/blob/a780a8d942716a9c6adccf0501c0f7a949f64211/Data/HashMap/Internal.hs#L1626-L1647

I expect that keeping `i` as a loop argument will be more efficient than recomputing it on each iteration though.

Reply:
Using `sparseIndex` makes benchmarks slower

Reply:
Can you show me the diff? 

> Using sparseIndex makes benchmarks slower

By how much? I also think the benchmark data might be a bit weird.

Reply:
How do I show you the diff?

Reply:
Maybe just push it, and push a revert if you think that's better.

Reply:
> It's slower by 10 microseconds

Can you paste the whole output for this benchmark?

Reply:
You mean the one with `sparseIndex`?

Reply:
Ideally both.

Reply:
Sparse index:
```
All
  HashMap
    intersection
      Int:        OK (0.34s)
        52.1 μs ± 3.3 μs
      ByteString: OK (0.26s)
        62.6 μs ± 6.1 μs
```

Reply:
Without sparse index:
```
All
  HashMap
    intersection
      Int:        OK (0.86s)
        42.1 μs ± 1.5 μs
      ByteString: OK (0.33s)
        47.7 μs ± 2.8 μs
```

Reply:
Alright, thanks! I think we might get different results with data where there's less overlap between the two maps. But that can be investigated at a different time.

Reply:
Follow-up issue in https://github.com/haskell-unordered-containers/unordered-containers/issues/416.